tools/Google.Cloud.Tools.SourceManipulation/SourceFile.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                throw new Exception($"Cannot remove a field from a declaration with multiple declarators.");
+            }
+            return WithNodeRemoved(matchingFields[0]);
+        }
+
+        /// <summary>
+        /// Rewrites the specified type (or namespace) name in a new form.
+        /// </summary>
+        /// <param name="oldTypeName">The old type name; may include an alias (e.g. foo::TypeName)</param>
+        /// <param name="newTypeName">The new type name; may include an alias (e.g. foo::TypeName)</param>
+        /// <returns>The modified the source file.</returns>
+        public SourceFile RewriteTypeName(string oldTypeName, string newTypeName) =>
+            Rewrite(new TypeNameRewriter(oldTypeName, newTypeName));
+
+        /// <summary>
+        /// Rewrites the specified type (or namespace) name in a new form.
+        /// </summary>
+        /// <param name="oldTypeName">The old type name; may include an alias (e.g. foo::TypeName)</param>
+        /// <param name="newTypeName">The new type name; may include an alias (e.g. foo::TypeName)</param>
+        /// <returns>The modified the source file.</returns>
+        public SourceFile RewriteMemberAccess(string oldAccessExpression, string newAccessExpression) =>
+            Rewrite(new MemberAccessRewriter(oldAccessExpression, newAccessExpression));
+
+        /// <summary>
+        /// Rewrites the source code using the given rewriter.
+        /// </summary>
+        /// <returns>The modified the source file.</returns>
+        public SourceFile Rewrite(CSharpSyntaxRewriter rewriter)
+        {
+            var newRoot = rewriter.Visit(_root);
+            // If we didn't find anything to change, that's almost always due to a bug, so fail quickly.
+            if (newRoot == _root)
+            {
+                throw new Exception("Rewriter made no changes");
+            }
+            return WithRootNode(newRoot);
+        }
+
+        /// <summary>
+        /// Returns the source file with a new root syntax node
+        /// </summary>
+        /// <returns>The modified the source file.</returns>
+        public SourceFile WithRootNode(SyntaxNode newRoot) =>
+            new SourceFile(_path, newRoot);
+
+        /// <summary>
+        /// Returns the source file with the given syntax node removed.
+        /// </summary>
+        /// <returns>The modified the source file.</returns>
+        public SourceFile WithNodeRemoved(SyntaxNode nodeToRemove) =>
+            WithRootNode(_root.RemoveNode(nodeToRemove, SyntaxRemoveOptions.KeepDirectives));
+
+        private class TypeNameRewriter : CSharpSyntaxRewriter
+        {
+            private readonly string _oldTypeName;
+            private readonly string _newTypeName;
+
+            public TypeNameRewriter(string oldTypeName, string newTypeName)
+                : base(visitIntoStructuredTrivia: true)
+            {
+                _oldTypeName = oldTypeName;
```

Comment:
nit: Why not `=> (_oldTypeName, _newTypeName) = ...` ??

Reply:
Could do... I tend to avoid doing that in this repo where a lot depends on versions without tuples. Will leave for now; we can always change that later.